src/geocoding/tests/test_geocoding.py

```diff
@@ -68,8 +68,8 @@ def test_entity_add_point_negative_coord(air_quality_observed):
     assert r['location']['type'] == 'geo:point'
 
     lon, lat = r['location']['value'].split(',')
-    assert float(lon) == pytest.approx(19.6389474, abs=1e-2)
-    assert float(lat) == pytest.approx(-98.9109537, abs=1e-2)
+    assert float(lon) == pytest.approx(19.5115112, abs=1e-2)
```

Comment:
was this one failing or this is just an adjustment of precision? If the former, we should go for the "at least one of many is correct" approach

Reply:
do we actually need to test this? or better should the test fail? IMHO, this is like "testing" the geo-decoding service being correct, which is out of our scope (I would say).

Reply:
we don't need to test the exact value used by OSM, but we need to test we couple well with whatever response they give, at least "in most cases" (hence my comment above).

That's why we don't need much precision in the test, just enough to recognise we're at least not mixing lat and long, things like that.

Reply:
understood and agreed, so purpose being to test if we support correctly returned: point, line, polygon, right? but still, if the change something for a search, we will end up in trouble. can we do a mockup test?

Reply:
i am thinking how to avoid that a change in the data of geocoding service screw up our integration tests

Reply:
> can we do a mockup test?

unfortunately mocking the response would not work because a "structural change" from their side would not make our test fail.

> i am thinking how to avoid that a change in the data of geocoding service screw up our integration tests

This is the key question. My suggestion discussed with @c0c0n3 was to choose say 5 points of big reliable cities and pass the test as long as at least one of them is correct (with gross tolerance, not being picky there). 

I didn't quite get yet why "picking the most accurate result" would solve the real issue (hence my question to @c0c0n3) :)

Reply:
yep, I fully agree with you guys on this: we should test structure rather than semantics---see my comment above. For example, I fixed the other test just by checking that what we get back is a polygon, rather than checking it is the polygon enclosing Mexico City. The reason I didn't do it for this too is that I'd like to check if my theory is correct. Let me explain what I have in mind.

If you look at the changes I made in the geocoding module, you'll see we're now picking the best match in terms of coordinates accuracy. I think these tests were breaking because of OSM returning *every time* the same result set for our geo query but with entries in a *different* order as in

    1. where is P? It's at: [(X, with accuracy ~ 0.1), (Y, with accuracy ~ 0.2)]
    2. where is P? It's at: [(Y, with accuracy ~ 0.2), (X, with accuracy ~ 0.1)]

If I'm right about this, then this test shouldn't fail going forward. If it fails again, it's probably a case of data corruption on the OSM side.

In any case, going forward I suggest we do the following:

1. Define what is the interface between us and a map provider---the `geocoder` lib isn't quite clear about what such an interface should be in terms of data being exchanged.
2. Provide a component implementing that interface by sourcing data from OSM.
3. Make the rest of our code work against that interface.
4. Use a mocked version of that interface for our unit tests.
5. Write *separate* integration tests to verify OSM works as (2) expects---this would catch OSM breaking API changes.

We basically have 2 and 3 in place. We need a bit more work for the rest which is why I haven't done it as part of this PR. Also, I've been thinking about test case generation using an approach similar to this: https://www.well-typed.com/blog/2019/01/qsm-in-depth/

But I suggest we talk about all this next time meet.

Reply:
Implemented relaxed coords checks as discussed earlier---see d98149b.